File: swift/Sources/Tennis/TennisGame2.swift
import Foundation

class TennisGame2: TennisGame {
    private let player1Name: String
    private let player2Name: String
    private var P1point: Int = 0
    private var P1res: String = ""
    private var P2point: Int = 0
    private var P2res: String = ""

    required init(player1: String, player2: String) {
        player1Name = player1
        player2Name = player2
    }

    var score: String? {
        var score = ""
        if P1point == P2point && P1point < 3
        {
            if P1point==0
            { score = "Love" }
            if P1point==1
            { score = "Fifteen" }
            if P1point==2
            { score = "Thirty" }
            score = "\(score)-All"
        }
        if P1point==P2point && P1point>2
        { score = "Deuce" }

        if P1point > 0 && P2point==0
        {
            if (P1point==1)
            { P1res = "Fifteen" }
            if (P1point==2)
            { P1res = "Thirty" }
            if (P1point==3)
            { P1res = "Forty" }
            
            P2res = "Love"
            score = "\(P1res)-\(P2res)"
        }
        if P2point > 0 && P1point==0
        {
            if (P2point==1)
            { P2res = "Fifteen" }
            if (P2point==2)
            { P2res = "Thirty" }
            if (P2point==3)
            { P2res = "Forty" }
            
            P1res = "Love"
            score = "\(P1res)-\(P2res)"
        }
        
        if (P1point>P2point && P1point < 4)
        {
            if (P1point==2)
            { P1res="Thirty" }
            if (P1point==3)
            { P1res="Forty" }
            if (P2point==1)
            { P2res="Fifteen" }
            if (P2point==2)
            { P2res="Thirty" }
            score = "\(P1res)-\(P2res)"
        }
        if P2point>P1point && P2point < 4
        {
            if (P2point==2)
            { P2res="Thirty" }
            if (P2point==3)
            { P2res="Forty" }
            if (P1point==1)
            { P1res="Fifteen" }
            if (P1point==2)
            { P1res="Thirty" }
            score = "\(P1res)-\(P2res)"
        }
        
        if P1point > P2point && P2point >= 3
        {
            score = "Advantage player1"
        }
        
        if P2point > P1point && P1point >= 3
        {
            score = "Advantage player2"
        }
        
        if P1point>=4 && P2point>=0 && (P1point-P2point)>=2
        {
            score = "Win for player1"
        }
        if P2point>=4 && P1point>=0 && (P2point-P1point)>=2
        {
            score = "Win for player2"
        }
        return score
    }
    
   private func setP1Score(number: Int) {
        
        for _ in 0..<number {
            P1Score()
        }
        
    }

    private func setP2Score(number: Int) {
        
        for _ in 0..<number {
            P2Score()
        }
        
    }
    
    private func P1Score() {
        P1point+=1
    }
    
    private func P2Score() {
        P2point+=1
    }
    
    func wonPoint(_ playerName: String) {
        if playerName == "player1" {
            P1Score()
        } else {
            P2Score()
        }
    }
}

File: c/TennisGame2.c
#include <stdlib.h>
#include <string.h>

#include "TennisGame2.h"

struct TennisGame
{
    int P1point;
    int P2point;
    const char* player1Name;
    const char* player2Name;
    const char* P1res;
    const char* P2res;
    char score[18];
};

struct TennisGame* TennisGame_Create(const char* player1Name, const char* player2Name)
{
    struct TennisGame* newGame = malloc(sizeof(struct TennisGame));
    newGame->P1point = 0;
    newGame->P2point = 0;
    newGame->P1res = "";
    newGame->P2res = "";
    newGame->player1Name = player1Name;
    newGame->player2Name = player2Name;
    return newGame;
}

const char* TennisGame_GetScore(struct TennisGame* game)
{
    game->score[0] = '\0';
    if (game->P1point == game->P2point && game->P1point < 4)
    {
        if (game->P1point == 0)
            strcpy(game->score, "Love");
        if (game->P1point == 1)
            strcpy(game->score, "Fifteen");
        if (game->P1point == 2)
            strcpy(game->score, "Thirty");
        strcat(game->score, "-All");
    }
    if (game->P1point == game->P2point && game->P1point > 2)
        strcpy(game->score, "Deuce");

    if (game->P1point > 0 && game->P2point == 0)
    {
        if (game->P1point == 1)
            game->P1res = "Fifteen";
        if (game->P1point == 2)
            game->P1res = "Thirty";
        if (game->P1point == 3)
            game->P1res = "Forty";

        game->P2res = "Love";
        strcat(strcat(strcpy(game->score, game->P1res), "-"), game->P2res);
    }
    if (game->P2point > 0 && game->P1point == 0)
    {
        if (game->P2point == 1)
            game->P2res = "Fifteen";
        if (game->P2point == 2)
            game->P2res = "Thirty";
        if (game->P2point == 3)
            game->P2res = "Forty";

        game->P1res = "Love";
        strcat(strcat(strcpy(game->score, game->P1res), "-"), game->P2res);
    }

    if (game->P1point > game->P2point && game->P1point < 4)
    {
        if (game->P1point == 2)
            game->P1res = "Thirty";
        if (game->P1point == 3)
            game->P1res = "Forty";
        if (game->P2point == 1)
            game->P2res = "Fifteen";
        if (game->P2point == 2)
            game->P2res = "Thirty";
        strcat(strcat(strcpy(game->score, game->P1res), "-"), game->P2res);
    }
    if (game->P2point > game->P1point && game->P2point < 4)
    {
        if (game->P2point == 2)
            game->P2res = "Thirty";
        if (game->P2point == 3)
            game->P2res = "Forty";
        if (game->P1point == 1)
            game->P1res = "Fifteen";
        if (game->P1point == 2)
            game->P1res = "Thirty";
        strcat(strcat(strcpy(game->score, game->P1res), "-"), game->P2res);
    }

    if (game->P1point > game->P2point && game->P2point >= 3)
    {
        strcpy(game->score, "Advantage player1");
    }

    if (game->P2point > game->P1point && game->P1point >= 3)
    {
        strcpy(game->score, "Advantage player2");
    }

    if (game->P1point >= 4 && game->P2point >= 0 && (game->P1point - game->P2point) >= 2)
    {
        strcpy(game->score, "Win for player1");
    }
    if (game->P2point >= 4 && game->P1point >= 0 && (game->P2point - game->P1point) >= 2)
    {
        strcpy(game->score, "Win for player2");
    }
    return game->score;
}

void TennisGame_SetP1Score(struct TennisGame* game, int number)
{
    for (int i = 0; i < number; i++)
    {
        TennisGame_P1Score(game);
    }
}

void TennisGame_SetP2Score(struct TennisGame* game, int number)
{
    for (int i = 0; i < number; i++)
    {
        TennisGame_P2Score(game);
    }
}

void TennisGame_P1Score(struct TennisGame* game)
{
    game->P1point++;
}

void TennisGame_P2Score(struct TennisGame* game)
{
    game->P2point++;
}

void TennisGame_WonPoint(struct TennisGame* game, const char* player)
{
    if (strcmp(player, "player1") == 0)
        TennisGame_P1Score(game);
    else
        TennisGame_P2Score(game);
}
